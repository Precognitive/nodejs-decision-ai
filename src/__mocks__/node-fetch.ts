// from http://www.wheresrhys.co.uk/fetch-mock/

import fetchMockModule from 'fetch-mock';

const nodeFetch = jest.requireActual('node-fetch');
const fetchMock = fetchMockModule.sandbox();
Object.assign(fetchMock.config, nodeFetch, {
  fetch: nodeFetch
});
export default fetchMock;
