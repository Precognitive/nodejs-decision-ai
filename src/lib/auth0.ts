// Auth0 Types, Interfaces and other structures
// ---------------------------------------------------


/**
 * @description Auth0 authentication protocol potential values
 * @link https://auth0.com/docs/rules/references/context-object
 */
const enum ContextProtocol {
  OidcBasicProfile = 'oidc-basic-profile',
  OidcImplicitProfile = 'oidc-implicit-profile',
  OAuth2ResourceOwner = 'oauth2-resource-owner',
  OAuth2ResourceOwnerJwtBearer = 'oauth2-resource-owner-jwt-bearer',
  OAuth2Password = 'oauth2-password',
  OAuth2RefreshToken = 'oauth2-refresh-token',
  SAMLP = 'samlp',
  WSFed = 'wsfed',
  WSTrustUsernameMixed = 'wstrust-usernamemixed',
  Delegation = 'delegation',
  RedirectCallback = 'redirect-callback',
}

/**
 * @description Auth0 Context Object, passed into rules
 * @link https://auth0.com/docs/rules/references/context-object
 */
interface Context {
  clientID: string,
  protocol: ContextProtocol,
  sessionID: string,
  request: {
    hostname: string,
    ip: string,
    query: string,
    userAgent: string,
    geoip: {
      city_name: string,
      continent_code: string,
      country_code3: string,
      country_code: string,
      country_name: string,
      latitude: string,
      longitude: string,
      time_zone: string,
    }
  }
}

/**
 * @description Auth0 User Object, passed into rules
 * @link https://auth0.com/docs/rules/references/user-object
 */
interface User {
  app_metadata: object,
  created_at: Date,
  email: string,
  last_ip: string,
  last_login: Date,
  logins_count: number,
  last_password_reset: Date,
  name: string,
  password_set_date: Date,
  updated_at: Date,
  username: string,
  user_id: string,
  user_metadata: object
}

/**
 * @description Auth0 Callback passed into rules
 * @link: https://auth0.com/docs/rules#syntax
 */
interface Callback {
  (err: null | Error, user: User, context: Context): void
}

/**
 * @description Auth0 Rule Interface
 * @link https://auth0.com/docs/rules
 */
interface Rule {
  (user: User, context: Context, callback: Callback): void
}

export { Context, ContextProtocol, User, Callback, Rule };
