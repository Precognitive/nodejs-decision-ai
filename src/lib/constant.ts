export const SDK_NAME = 'Cognition';
