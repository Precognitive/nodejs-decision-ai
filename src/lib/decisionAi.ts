// Cognition SDK
// ---------------------------------------------------

const enum ApiVersion {
  v1 = 'v1'
}

const enum DecisionStatus {
  allow = 'allow',
  review = 'review',
  reject = 'reject'
}

const enum Channel {
  web = 'web',
  desktop = 'desktop',
  app = 'app',
}

const enum LoginStatus {
  success = 'success',
  failure = 'failure'
}

const enum AuthenticationType {
  client_storage = 'client_storage',
  password = 'password',
  two_factor = 'two_factor',
  single_sign_on = 'single_sign_on',
  social_sign_on = 'social_sign_on',
  key = 'key'
}

interface CognitionResponse {
  score: number,
  confidence: number,
  decision: DecisionStatus,
  tokenId: string,
  signals: Array<string>
}

interface CognitionInput extends CognitionRequestOverrides {
  eventId?: string;
  dateTime?: Date;
  ipAddress: string;
  login: {
    userId: string;
    channel: Channel;
    usedCaptcha: boolean;
    usedRememberMe?: boolean;
    authenticationType?: AuthenticationType | null;
    status: LoginStatus;
    passwordUpdateTime: Date;
    userNameUpdateTime?: Date;
  };
}
type CognitionRequestOverrides = Readonly<CognitionInput>;

interface CognitionRequest extends CognitionInput {
  apiKey: string;
  dateTime: Date;
}

export {
  ApiVersion,
  DecisionStatus,
  Channel,
  LoginStatus,
  AuthenticationType,
  CognitionInput,
  CognitionResponse,
  CognitionRequest,
  CognitionRequestOverrides
};
