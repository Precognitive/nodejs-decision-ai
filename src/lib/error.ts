import { SDK_NAME } from './constant';

class DecisionError extends Error {
  public readonly isFraudulent: boolean;

  constructor(isFraudulent = true) {
    super('Precognitive: Reject Authentication');
    this.isFraudulent = isFraudulent;
  }
}

class HttpError extends Error {
  public readonly statusCode: number;
  public readonly response?: any;
  public readonly body?: any;

  constructor(statusCode: number, response?: any, body?: any) {
    super(`${SDK_NAME} - HTTP Error [${statusCode}]`);

    this.statusCode = statusCode;
    this.response = response;
    this.body = body;
  }
}

export {DecisionError, HttpError};
