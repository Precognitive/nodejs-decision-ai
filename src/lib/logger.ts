import { SDK_NAME } from './constant';

const enum LogLevel {
  DEBUG = 4,
  INFO = 3,
  WARN = 2,
  ERROR = 1,
  NONE = 0
}

class ConsoleLogger implements Logger {
  private readonly logLevel: LogLevel;

  constructor(logLevel: LogLevel) {
    this.logLevel = logLevel;
  }

  public debug(...args: any) {
    if (this.logLevel >= LogLevel.DEBUG) {
      console.debug(`${SDK_NAME} DEBUG:`, ...args);
    }
  }

  public info(...args: any) {
    if (this.logLevel >= LogLevel.INFO) {
      console.info(`${SDK_NAME} INFO:`, ...args);
    }
  }

  public warn(...args: any) {
    if (this.logLevel >= LogLevel.WARN) {
      console.warn(`${SDK_NAME} WARN:`, ...args);
    }
  }

  public error(...args: any) {
    if (this.logLevel >= LogLevel.ERROR) {
      console.error(`${SDK_NAME} ERROR:`, ...args);
    }
  }
}

interface Logger {
  debug(...args: any): void;
  info(...args: any): void;
  warn(...args: any): void;
  error(...args: any): void;
}

export {ConsoleLogger, LogLevel, Logger};
