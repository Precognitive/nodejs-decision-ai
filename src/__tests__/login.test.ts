import fetchMockModule from 'fetch-mock';
import _ from 'lodash';
import fetchMock from '../__mocks__/node-fetch';
import { Login, DecisionError, HttpError } from '../';
import { ApiVersion, CognitionResponse, DecisionStatus, AuthenticationType, Channel, LoginStatus, CognitionInput } from '../lib/decisionAi';
import { RequestInit } from 'node-fetch';

const date = new Date();

// from https://stackoverflow.com/questions/3143070/javascript-regex-iso-datetime
const isoDateRegexp = /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/;

function getBaseResponse (decision: DecisionStatus, score = _.random(-200, 200), confidence = _.random(0, 100)): CognitionResponse {
  return {
    decision,
    score,
    confidence,
    signals: [
      'test',
      'test2'
    ]
  }
}

const config = {
  auth: {
    userName: 'test',
    password: 'test-pw'
  },
  apiKey: 'kasf8w-afsafs-1asffw',
  version: ApiVersion.v1
};
const basicAuth = 'Basic dGVzdDp0ZXN0LXB3';

function getOptions (): CognitionInput {
  return {
    eventId: '123456789',
    ipAddress: '127.0.0.1',
    login: {
      authenticationType: AuthenticationType.password,
      channel: Channel.web,
      passwordUpdateTime: date,
      status: LoginStatus.success,
      usedCaptcha: false,
      userId: '34294892831981',
    },
  };
}

function getDecisionRecord () {
  return {
    apiKey: config.apiKey,
    dateTime: expect.stringMatching(isoDateRegexp),
    eventId: '123456789',
    ipAddress: '127.0.0.1',
    login: {
      authenticationType: 'password',
      channel: 'web',
      passwordUpdateTime: date.toISOString(),
      status: 'success',
      usedCaptcha: false,
      userId: '34294892831981',
    },
  };
}

function assertOneCall(calls: fetchMockModule.MockCall[], body: any) {
  expect(calls.length).toEqual(1);

  // the fetch mock api is asinine
  // it adds a 'request' property to the array that we don't care about
  // this strips the property and turns it into a real array.
  const request = [...calls[0]];

  expect(request).toEqual([url, {
    body: expect.any(String),
    headers: {
      Authorization: basicAuth
    },
    method: 'POST',
    timeout: 5000
  }]);
  expect(JSON.parse((request[1] as RequestInit).body as any)).toEqual(body);
}

const url = 'https://api.precognitive.io/v1/decision/login';

afterEach(fetchMock.restore);

describe('decision', () => {
  it('returns decision-ai response', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const response = getBaseResponse(DecisionStatus.allow);
    fetchMock.postOnce(url, response);
    const actual = await login.decision(defaultOptions);
    expect(actual).toEqual(response);

    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('throws error on http failure', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    fetchMock.postOnce(url, {status: 500, body: JSON.stringify({code: 'failed', message: 'Failed'})});
    await expect(login.decision(defaultOptions)).rejects.toBeInstanceOf(HttpError);
    expect(fetchMock.done()).toEqual(true);
  });

  it('throws error on request failure', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const err = new Error('Mock');
    fetchMock.postOnce(url, {throws: err});
    await expect(login.decision(defaultOptions)).rejects.toBe(err);
    expect(fetchMock.done()).toEqual(true);
  });
});

describe('autodecision', () => {
  it('disallows if the response is `reject`', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const response = getBaseResponse(DecisionStatus.reject);
    fetchMock.postOnce(url, response);
    await expect(login.autoDecision(defaultOptions)).rejects.toBeInstanceOf(DecisionError);

    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows if the response is `review`', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const response = getBaseResponse(DecisionStatus.review);
    fetchMock.postOnce(url, response);
    await login.autoDecision(defaultOptions);

    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows if the response is `allow`', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const response = getBaseResponse(DecisionStatus.allow);
    fetchMock.postOnce(url, response);
    login.autoDecision(defaultOptions);

    assertOneCall(fetchMock.calls(), getDecisionRecord());
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows on http failure', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    fetchMock.postOnce(url, {status: 500, body: JSON.stringify({code: 'failed', message: 'Failed'})});
    await login.autoDecision(defaultOptions);
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows on request failure', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const err = new Error('Mock');
    fetchMock.postOnce(url, {throws: err});
    await login.autoDecision(defaultOptions);
    expect(fetchMock.done()).toEqual(true);
  });

  it('allows default override on request failure', async () => {
    const defaultOptions = getOptions();
    const login = new Login(config);
    const err = new Error('Mock');
    fetchMock.postOnce(url, {throws: err});
    await expect(login.autoDecision(defaultOptions, { defaultResponse: DecisionStatus.reject })).rejects.toBeInstanceOf(DecisionError);
    expect(fetchMock.done()).toEqual(true);
  });
});

describe('isGoodLogin', () => {
  it('accepts allowed login', () => {
    expect(Login.isGoodLogin(getBaseResponse(DecisionStatus.allow))).toEqual(true);
  });

  it('accepts reviewed login', () => {
    expect(Login.isGoodLogin(getBaseResponse(DecisionStatus.review))).toEqual(true);
  });

  it('rejects rejected login', () => {
    expect(Login.isGoodLogin(getBaseResponse(DecisionStatus.reject))).toEqual(false);
  });
});

it('uses custom logger', async () => {
  const logger = {
    debug: jest.fn(),
    info: jest.fn(),
    warn: jest.fn(),
    error: jest.fn()
  };
  const defaultOptions = getOptions();
  defaultOptions.dateTime = new Date('2019-01-01');
  defaultOptions.login.passwordUpdateTime = defaultOptions.dateTime;
  const login = new Login({...config, logger});
  const response = getBaseResponse(DecisionStatus.allow, 50, 40);
  fetchMock.postOnce(url, response);
  await login.decision(defaultOptions);
  expect(_.mapValues(logger, (mock) => mock.mock.calls)).toMatchSnapshot();
  expect(fetchMock.done()).toEqual(true);
});
