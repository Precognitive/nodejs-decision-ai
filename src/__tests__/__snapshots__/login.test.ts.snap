// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uses custom logger 1`] = `
Object {
  "debug": Array [
    Array [
      "REQUEST BODY - {\\"dateTime\\":\\"2019-01-01T00:00:00.000Z\\",\\"eventId\\":\\"123456789\\",\\"ipAddress\\":\\"127.0.0.1\\",\\"login\\":{\\"authenticationType\\":\\"password\\",\\"channel\\":\\"web\\",\\"passwordUpdateTime\\":\\"2019-01-01T00:00:00.000Z\\",\\"status\\":\\"success\\",\\"usedCaptcha\\":false,\\"userId\\":\\"34294892831981\\"},\\"apiKey\\":\\"kasf8w-afsafs-1asffw\\"}",
    ],
    Array [
      "RESPONSE BODY - {\\"decision\\":\\"allow\\",\\"score\\":50,\\"confidence\\":40,\\"token\\":\\"4610b4ba-a1e0-4f8c-8cc3-f5d9bd2214cc\\",\\"signals\\":[\\"test\\",\\"test2\\"]}",
    ],
  ],
  "error": Array [],
  "info": Array [],
  "warn": Array [],
}
`;
